de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/concurrent/management/JupiterClient.java

```diff
@@ -7,34 +7,49 @@
 import de.fu_berlin.inf.dpp.concurrent.jupiter.Operation;
 import de.fu_berlin.inf.dpp.concurrent.jupiter.TransformationException;
 import de.fu_berlin.inf.dpp.concurrent.jupiter.internal.Jupiter;
+import de.fu_berlin.inf.dpp.concurrent.jupiter.internal.text.TimestampOperation;
+import de.fu_berlin.inf.dpp.session.AbstractActivityProducer;
 import de.fu_berlin.inf.dpp.session.ISarosSession;
-import java.util.HashMap;
+import de.fu_berlin.inf.dpp.util.NamedThreadFactory;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 
 /** A JupiterClient manages Jupiter client docs for a single user with several paths */
-public class JupiterClient {
-
-  protected ISarosSession sarosSession;
-
-  public JupiterClient(ISarosSession sarosSession) {
-    this.sarosSession = sarosSession;
-  }
+public class JupiterClient extends AbstractActivityProducer {
 
+  private final ISarosSession sarosSession;
   /**
    * Jupiter instances for each local editor.
    *
    * @host and @client
    */
-  protected final HashMap<SPath, Jupiter> clientDocs = new HashMap<SPath, Jupiter>();
+  private final Map<SPath, Jupiter> clientDocs = new ConcurrentHashMap<>();
+
+  private final ScheduledThreadPoolExecutor jupiterHeartbeat =
+      new ScheduledThreadPoolExecutor(1, new NamedThreadFactory("dpp-jupiter-heartbeat"));
+
+  public JupiterClient(ISarosSession sarosSession) {
+    this.sarosSession = sarosSession;
+    jupiterHeartbeat.scheduleWithFixedDelay(
```

Comment:
Beside this lambda madness I do not understand the reason to put the logic into this class. First the server/host also have a running instance of this class. Second, do we have to use TimestampOperations or is a NOP operation sufficient ?

As for the thread access. While you are allowed to call fireActivity from any thread you cannot access Jupiter client stuff from any thread. This has to be done in the UI / Application thread through the UISynchronizer. I do not prefer a solution with monitors or other concurrent data types. It only complicates the code. There are already enough synchronized stuff in here that should not exists.